### PR TITLE
Reimagined: Correctly handle lengths of serialized strings

### LINKED ISSUE #33: Incorrect length calculation in serialization of string data
In the attached SQL dumps of a WordPress options table, some string data is not being serialized correctly.

Command: ```cat wp_1000642_options.sql | ./go-search-replace hcommons.org heinlein.mlacommons.org > wp_1000642_options_replaced.sql```

The error occurs for one long field containing CSS. Here are the relevant snippets:

Original:
```
...
s:16:\"additional-style\";s:24458:\"div.gdlr-blog-co
...
```

There is one substitution performed of hcommons.org -> heinlein.mlacommons.org

Result:
```
...
s:16:\"additional-style\";s:6533:\"div.gdlr-blog-co
...
```

The new serialized length should be 24469, reflecting the difference in string length between the two URLs, but it is instead reported as 6533. If I serialize the value using PHP's serialize function, it is serialized correctly. The serialization appears correct other than the length.

Original file (renamed .txt for uploading):
[wp_1000642_options.txt](https://github.com/Automattic/go-search-replace/files/7339962/wp_1000642_options.txt)

File after replacement:
[wp_1000642_options_replaced.txt](https://github.com/Automattic/go-search-replace/files/7339959/wp_1000642_options_replaced.txt)

### PR DESCRIPTION
**Pull request by abdullah-kasim**
Hello, this PR is an alternative to the PR in #39 and should fix #33 .

Regex isn't too great for this purpose, as it has a tendency to over-match. Hence, this PR offers an alternative solution.

This PR uses a mini-tokenizer/parser to detect instances of PHP String, then do search-replace in the constraints of the token, then go back up the stack to recalculate the new byte size.

Special care has been taken care so that we didn't include MySQL's sqldump escaping.

It should technically work with mydumper as well - upon inspecting a mydumper dump, the escaping is very much the same.

There's a few errors with the tests in the PR #39 which I've fixed, as well as added new tests to handle some edge cases with excessive escaping, as well as with very plausible scenario of someone using the string combination of `\";` i.e. in WP posts.

If approved, I'll re-point the branch to trunk and merge it from there.

### CODE REVIEW

File: main_test.go [hunk 1]
```diff
@@ -91,8 +91,8 @@ func TestSerializedReplaceWithCss(t *testing.T) {
 		"https://ncc-1701-d.space",
 	}
 
-	input := `a:2:{s:3:\"key\";s:5:\"value\";s:3:\"css\";s:216:\"body { color: #123456;\r\nborder-bottom: none; }\r\ndiv.bg { background: url('https://uss-enterprise.com/wp-content/uploads/main-bg.gif');\r\n  background-position: left center;\r\n    background-repeat: no-repeat; }\";}`
-	expected := `a:2:{s:3:\"key\";s:5:\"value\";s:3:\"css\";s:214:\"body { color: #123456;\r\nborder-bottom: none; }\r\ndiv.bg { background: url('https://ncc-1701-d.space/wp-content/uploads/main-bg.gif');\r\n  background-position: left center;\r\n    background-repeat: no-repeat; }\";}`
+	input := `a:2:{s:3:\"key\";s:5:\"value\";s:3:\"css\";s:208:\"body { color: #123456;\r\nborder-bottom: none; }\r\ndiv.bg { background: url('https://uss-enterprise.com/wp-content/uploads/main-bg.gif');\r\n  background-position: left center;\r\n    background-repeat: no-repeat; }\";}`
+	expected := `a:2:{s:3:\"key\";s:5:\"value\";s:3:\"css\";s:206:\"body { color: #123456;\r\nborder-bottom: none; }\r\ndiv.bg { background: url('https://ncc-1701-d.space/wp-content/uploads/main-bg.gif');\r\n  background-position: left center;\r\n    background-repeat: no-repeat; }\";}`
```
Comment (abdullah-kasim) [inline main_test.go:95]: I used a PHP REPL and realized that these byte lengths are not correct - I've added the correct byte length

File: main_test.go [hunk 2]
```diff
@@ -102,7 +102,7 @@ func TestSerializedReplaceWithCssAndUnrelatedSerializationMarker(t *testing.T) {
...
-	expected := `a:2:{s:3:\"key\";s:5:\"value\";s:3:\"css\";s:247:\"body { color: #123456;\r\nborder-bottom: none; }\r\nbody:after{ content: \"▼\"; }\r\ndiv.bg { background: url('https://ncc-1701-d.space/wp-content/uploads/main-bg.gif');\r\n  background-position: left center;\r\n    background-repeat: no-repeat; }\";}`
+	input := `a:2:{s:3:\"key\";s:5:\"value\";s:3:\"css\";s:239:\"body { color: #123456;\r\nborder-bottom: none; }\r\nbody:after{ content: \"▼\"; }\r\ndiv.bg { background: url('https://uss-enterprise.com/wp-content/uploads/main-bg.gif');\r\n  background-position: left center;\r\n    background-repeat: no-repeat; }\";}`
+	expected := `a:2:{s:3:\"key\";s:5:\"value\";s:3:\"css\";s:237:\"body { color: #123456;\r\nborder-bottom: none; }\r\nbody:after{ content: \"▼\"; }\r\ndiv.bg { background: url('https://ncc-1701-d.space/wp-content/uploads/main-bg.gif');\r\n  background-position: left center;\r\n    background-repeat: no-repeat; }\";}`
```
Comment (abdullah-kasim) [inline main_test.go:106]: Ditto, incorrect byte length.

File: search-replace.go [hunk 3]
```diff
@@ -155,71 +162,229 @@ func fixLine(line *[]byte, replacements []*Replacement) *[]byte {
...
-		Debugf("Start of loop, startIndex: %d\n", startIndex)
-		match := serializedStringPrefixRegexp.FindSubmatchIndex((*line)[startIndex:])
-		if match == nil {
+	index := 0
+
+	var rebuiltLine []byte
+
+	for index < len(*line) {
+		Debugf("Start of loop, index: %d\n", index)
+		linePart := (*line)[index:]
```
Comment (abdullah-kasim) [inline search-replace.go]: Once a certain part of the line has been processed, we slice it away to create a new `linePart`

File: search-replace.go [hunk 4]
```diff
@@ -155,71 +162,229 @@ func fixLine(line *[]byte, replacements []*Replacement) *[]byte {
...
+
+		details, err := parseEscapedData(linePart)
```
Comment (abdullah-kasim) [inline search-replace.go]: We then figure out what's within that remaining `linePart` - i.e. where does the content start, the end of it, where the 'next part' should be. `NextPart` here refers to the point where we stop parsing the current part - and that's at the last instance of the string `\";` i.e. in `s:1:\"a\";`. The next part does not necessarily immediately start at another serialized string - `parseEscapedData` has a job to find where the starting point of the next serialized string.

File: search-replace.go [hunk 5]
```diff
@@ -155,71 +162,229 @@ func fixLine(line *[]byte, replacements []*Replacement) *[]byte {
...
+			contentLength = details.OriginalByteSize
 		}
-		Debugf("Match found, length: %d\n", length)
 
-		contentStart := startIndex + match[1]
-		contentEnd := contentStart + length
+		// and we rebuild the string
+		rebuilt := "s:" + strconv.Itoa(contentLength) + ":\\\"" + string(updatedContent) + "\\\";"
+
+		rebuiltLine = append(rebuiltLine, []byte(rebuilt)...)
```
Comment (abdullah-kasim) [inline search-replace.go]: Serialized value rebuilding, replacing, and escaping, and offsetting mostly. `unescapeContent` so that we can remove all the escape backslashes from mysqldump's serialization. We need this because the content length rely on the unescaped's byte size, rather than the escaped's byte size.

File: search-replace.go [hunk 6]
```diff
@@ -155,71 +162,229 @@ func fixLine(line *[]byte, replacements []*Replacement) *[]byte {
...
+}
+
+func replaceInSerializedBytes(serialized []byte, replacements []*Replacement) []byte {
+	for _, replacement := range replacements {
+		serialized = bytes.ReplaceAll(serialized, replacement.From, replacement.To)
+	}
+	return serialized
+}
+
+var serializedStringPrefixRegexp = regexp.MustCompile(`s:(\d+):`)
```
Comment (abdullah-kasim) [inline search-replace.go]:
```suggestion
var serializedStringPrefixRegexp = regexp.MustCompile(`s:(\d+):\\"`)
```
I'll fix this - I think I accidentally removed this.

File: search-replace.go [hunk 7]
```diff
@@ -155,71 +162,229 @@ func fixLine(line *[]byte, replacements []*Replacement) *[]byte {
...
+		ContentStartIndex: 0,
+		ContentEndIndex:   0,
+		NextPartIndex:     0,
+		CurrentPartIndex:  0,
+		OriginalByteSize:  0,
+	}
+
+	// find starting point in the line
+	//TODO: We should first check if we found the string when inside a quote or not.
+	// but currently skipping that scenario because it seems unlikely to find it outside.
```
Comment (abdullah-kasim) [inline search-replace.go]: Rather hard to do, but it's possible that the regex matches inside say, a JSON, and it'd have some sort of multi-layer escaping. Someone might also accidentally use the same combination of string i.e. `s:123:"`, as maybe he's writing a programming tutorial. So we'd accidentally match these, and we should match as little of these as possible, because it'll affect the parser's stability.

File: search-replace.go [hunk 8]
```diff
@@ -155,71 +162,229 @@ func fixLine(line *[]byte, replacements []*Replacement) *[]byte {
...
+	// 2. \\ in dump becomes \ when inserting a mysql row.
+	// 3. \' in dump becomes ' when inserting a mysql row.
+	// 4. mysql translates newline into \n when creating a mysqldump. Same applies to carriage return.
+	// 5. PHP serialize does not convert the bytes \r or \n into something else - they're as-is.
+	// 6. If using single quotes in php, \r and \n does not get converted into bytes - they become literal backslash and letter.
+	// Generally, to unescape, we need to do the following:
+	// 1. Convert \\ to \
+	// 2. Convert \' to '
+	// 3. Convert \" to "
+
```
Comment (abdullah-kasim) [inline search-replace.go]: I'll add a few more conditions to this PR here - we should be converting everything within `unescapedMap` just above. Not sure what else I missed.

File: search-replace.go [hunk 9]
```diff
@@ -155,71 +162,229 @@ func fixLine(line *[]byte, replacements []*Replacement) *[]byte {
...
+	}
+
+	actualByte := unescapedMap[charPair[1]]
+
+	if actualByte != 0 {
+		return []byte{actualByte}
+	}
+
+	// what if it's not a valid escape? Do nothing - it's considered as already escaped
+	return charPair
```
Comment (abdullah-kasim) [inline search-replace.go:308]: I need to double check if this is correct - admittedly I had GH Copilot fill this up.

File: main_test.go (hunk 1 above)
Comment (sanmai) [reply main_test.go:95]: I could confirm the previous strings were invalid: https://3v4l.org/DJVBo

**Review by luismulinari — COMMENTED** [review]
This is awesome!

Just wondering if it would be worth adding some test cases from: https://github.com/php/php-src/tree/master/ext/standard/tests/serialize

It has a pretty good set of edge cases.

**Comment by nickchomey** [comment]
Nice work! I was thinking about this problem recently and found these two packages for handling php serialized strings in Go.

* https://github.com/elliotchance/phpserialize
* https://github.com/trim21/go-phpserialize

I asked in an issue in the 2nd repo what the difference was and the dev gave a nice answer. 
https://github.com/trim21/go-phpserialize/issues/53

I wonder if this (finding the serialized strings, unserialize, modify, reserialize) would be a more robust way of handling this? (I brought it up in #42 but perhaps it wasn't noticed/relevant)

**Review by luismulinari — COMMENTED** [review]
Also, do you think it would be worth adding some [fuzzy tests](https://go.dev/doc/security/fuzz/)?

**Comment by abdullah-kasim** [comment]
Yeah I definitely need more tests. Thanks for those suggestions. @luismulinari !

**Comment by abdullah-kasim** [comment]
> Nice work! I was thinking about this problem recently and found these two packages for handling php serialized strings in Go.
> 
> * [elliotchance/phpserialize](https://github.com/elliotchance/phpserialize)
> * [trim21/go-phpserialize](https://github.com/trim21/go-phpserialize)
> 
> I asked in an issue in the 2nd repo what the difference was and the dev gave a nice answer. [trim21/go-phpserialize#53](https://github.com/trim21/go-phpserialize/issues/53)
> 
> I wonder if this (finding the serialized strings, unserialize, modify, reserialize) would be a more robust way of handling this? (I brought it up in #42 but perhaps it wasn't noticed/relevant)

@nickchomey I did notice your post about these libraries, and yeah I agree I should add some comments about it :) . The reason why I didn't utilize these libraries, is that it came down to simplification and speed.

Neither of these two libraries are designed to parse something from a sqldump, so we first need to parse the sqldump so that these serialization library can do its work.

Which means, either we write a SQL parser, or use something existing, like `vitess-sqlparser`. But we can't simply use a SQL parser, as [per this reason](https://github.com/blastrain/vitess-sqlparser#why).

> # Why

> [xwb1989/sqlparser](https://github.com/xwb1989/sqlparser) is famous sql parser in Go.  
But it cannot parse some query (like offset or bulk insert...) because it customizes vitess's sql parser.  

> Also, some libraries use from vitess sql parser directly. But vitess's sql parser only partial supports DDL parsing.  

> We want to perfectly support parsing for SQL and DDL.  
> Therefore we use vitess sql parser directly and also use TiDB parser for DDL parsing. 

And so we're hitting a complexity point here - if we use a pre-existing sql parser, we're at the whims of that parser library. Certain lines in the sqldump might not work. Maybe it doesn't support certain mysql syntax, maybe we have some nasty edge-case in the sqldump, we'd need to check for mydumper compatibility as well. Admittedly I only did some thought experiments but didn't test-drive it for real. And I think it'll be slower than this PR.

And that's how I decided to just write my own parser. Took at most a day for the core code.

**Comment by nickchomey** [comment]
Fair enough! Hopefully this can get merged soon enough!

**Comment by abdullah-kasim** [comment]
> This is awesome!
> 
> Just wondering if it would be worth adding some test cases from: [php/php-src@`master`/ext/standard/tests/serialize](https://github.com/php/php-src/tree/master/ext/standard/tests/serialize?rgh-link-date=2024-10-07T12%3A26%3A24Z)
> 
> It has a pretty good set of edge cases.

@luismulinari There's a bit too many test cases in there, and there's a lot that is unrelated to string serialization. Anything specific that you wish to point out?

File: search-replace.go [hunk 10]
```diff
@@ -155,71 +162,229 @@ func fixLine(line *[]byte, replacements []*Replacement) *[]byte {
...
+		'\\': '\\',
+		'\'': '\'',
+		'"':  '"',
+		'n':  '\n',
+		'r':  '\r',
+		't':  '\t',
+		'b':  '\b',
+		'f':  '\f',
+		'0':  '\x00',
+	}
```
Comment (abdullah-kasim) [inline search-replace.go:299]: List verified by running this query in a mysql db:

```sql
insert into string_test values ('\a\b\c\d\e\f\g\h\i\j\k\l\m\n\o\p\q\r\s\t\u\v\w\x\y\z\\\!\@\#\$\%\^\&\*\(\)\-\=\{\}\[\]\;\'\,\.\/\<\>\?\:\"\`\~\_\+\|\0\1\2\3\4\5\6\7\8\9');
```

Then generating a db dump from there - yields the following string.

```sql
'acdefghijklm\nopq\rs	uvwxyz\\!@#$\\%^&*()-={}[];\',./<>?:\"`~\\_+|\0123456789'
```

`\b` is translated into something in mysql, while `\t` is a tab as-is. Technically we could remove `\t` from the above list.

File: search-replace.go (hunk 9 above)
Comment (abdullah-kasim) [reply search-replace.go:308]: Verified, should be correct.

**Comment by abdullah-kasim** [comment]
> Also, do you think it would be worth adding some [fuzzy tests](https://go.dev/doc/security/fuzz/)?

@luismulinari 

Probably not worth it - rather difficult to fuzz due to escaping and all so the fuzzing needs to have escaping logic itself 😅 , and the function pretty obviously short-circuits to just returning the line if it fails to parse it. Not enough branching for it, I suppose.

At least that's my take on it right now.

**Comment by abdullah-kasim** [comment]
Benchmarks in case they are a concern:

```
cpu: Apple M2 Max
BenchmarkFix-12                          1864381               641.4 ns/op
BenchmarkNoReplaceOld-12                149462286                8.210 ns/op
BenchmarkNoReplaceNew-12                 6201714               193.3 ns/op
BenchmarkSimpleReplaceOld-12            100000000               10.71 ns/op
BenchmarkSimpleReplaceNew-12             5087581               238.7 ns/op
BenchmarkSerializedReplaceOld-12        70218040                17.00 ns/op
BenchmarkSerializedReplaceNew-12         4164714               286.2 ns/op
```

20x slow-down?

Real world: (2.8 GB DB dump)

```
[ 18:18:09 ] time cat dump.sql | ./go-search-replace-new domain-a.com domain-b.com > test-1.sql

real    0m2.093s
user    0m4.952s
sys     0m3.376s

go-search-replace on  trunk [📝] via  v1.23.2 via 🐘 v8.3.12 took 2s 
[ 18:18:14 ] time cat dump .sql | ./go-search-replace-old domain-a.com domain-b.com > test-1.sql

real    0m1.332s
user    0m1.661s
sys     0m1.818s
```

Not too bad I suppose...

File: search-replace_test.go [hunk 11]
```diff
@@ -93,11 +163,59 @@ func TestReplace(t *testing.T) {
...
+		},
+		{
+			testName: "search and replace with different lengths",
+
+			from: []byte("bbbbbbbbbb"),
+			to:   []byte("ccccccccccccccc"),
+
+			in:  []byte(`s:20:\"aaaaabbbbbbbbbbaaaaa\";`),
+			out: []byte(`s:25:\"aaaaacccccccccccccccaaaaa\";`),
+		},
```
Comment (luismulinari) [inline search-replace_test.go:226]: I've found a few edge cases on the php-src tests: https://github.com/php/php-src/tree/master/ext/standard/tests/serialize

Not sure if it's worth fixing them all.

```suggestion
		},
		{
			testName: "exotic letters",

			from: []byte("åäöÅÄÖ"),
			to:   []byte("---åäöÅÄÖ---"),

			in:  []byte(`a:1:{s:12:"åäöÅÄÖ";s:12:"åäöÅÄÖ";}`),
			out: []byte(`a:1:{s:18:"---åäöÅÄÖ---";s:18:"---åäöÅÄÖ---";}`),
		},
		{
			testName: "complex array",

			from: []byte("hallo"),
			to:   []byte("world!"),

			in:  []byte(`a:6:{i:0;i:1;i:1;d:1.1;i:2;s:5:"hallo";i:3;N;i:4;b:1;i:5;a:0:{}}`),
			out: []byte(`a:6:{i:0;i:1;i:1;d:1.1;i:2;s:6:"world!";i:3;N;i:4;b:1;i:5;a:0:{}}`),
		},
		{
			testName: "object",

			from: []byte("hello"),
			to:   []byte("world!"),

			in:  []byte(`O:8:"stdClass":1:{s:3:"key";s:5:"hello";}`),
			out: []byte(`O:8:"stdClass":1:{s:3:"key";s:6:"world!";}`),
		},
		{
			testName: "object with recursion",

			from: []byte("hello!"),
			to:   []byte("world!!!"),

			in:  []byte(`O:8:"stdClass":2:{s:4:"obj2";O:8:"stdClass":1:{s:4:"obj3";O:8:"stdClass":1:{s:4:"obj1";r:1;}}s:5:"hello";s:6:"hello!";}`),
			out: []byte(`O:8:"stdClass":2:{s:4:"obj2";O:8:"stdClass":1:{s:4:"obj3";O:8:"stdClass":1:{s:4:"obj1";r:1;}}s:5:"world";s:8:"world!!!";}`),
		},
		{
			testName: "replace object key with invalid value",

			from: []byte("valid_key"),
			to:   []byte("invalid_key!!"),

			in:  []byte(`O:8:"stdClass":1:{s:9:"valid_key";s:4:"data";}`),
			out: []byte(`O:8:"stdClass":1:{s:9:"valid_key";s:4:"data";}`), //Should we replace the key or not?
		},
		{
			testName: "object with replace in key and value",

			from: []byte("t1"),
			to:   []byte("t999"),

			in:  []byte(`O:8:"stdClass":1:{s:2:"t1";a:1:{s:2:"t1";a:1:{i:0;s:2:"t1";}}}`),
			out: []byte(`O:8:"stdClass":1:{s:4:"t999";a:1:{s:4:"t999";a:1:{i:0;s:4:"t999";}}}`),
		},		
```

Comment (abdullah-kasim) [reply search-replace_test.go:226]: Not really worth it. I pondered over it but ultimately decided not to. We're inherently using bytes for all matches and bytelength so the encoding doesn't really matter.

a 2-byte utf-8 character, when encoded to ascii only for example... is still 2 bytes - just that it's now 2 characters. We already have tests for emojis so that's a suitable equivalent for accented letters.

Comment (abdullah-kasim) [reply search-replace_test.go:226]: As for the object keys... probably not worth it, and we might want to replace the keys as well, too. Think of key-value where the key is a domain.

Comment (abdullah-kasim) [reply search-replace_test.go:226]: I'll double check these tests though - as I'm sure the ones with accented characters should work.

Comment (abdullah-kasim) [reply search-replace_test.go:226]: ~Okay, yes, worth fixing. Thanks for these @luismulinari~

~Kinda just assumed that it's working~

Comment (abdullah-kasim) [reply search-replace_test.go:226]: Just double-checked the tests - You need to escape it the same way as mysql escapes it i.e. put it in a db, and make a mysqldump.

The search-replace serializer does not apply to non-sql-dumps to keep the complexity low. We might want to consider it later once we've seen usages related to it.

Comment (abdullah-kasim) [reply search-replace_test.go:226]: i.e. instead of this:
```golang
		{
			testName: "exotic letters",
			from:     []byte("åäöÅÄÖ"),
			to:       []byte("---åäöÅÄÖ---"),
			in:       []byte(`a:1:{s:12:"åäöÅÄÖ";s:12:"åäöÅÄÖ";}`),
			out:      []byte(`a:1:{s:18:"---åäöÅÄÖ---";s:18:"---åäöÅÄÖ---";}`),
		},
```


Do this (notice the escaped quotes)

```golang
		{
			testName: "exotic letters",
			from:     []byte("åäöÅÄÖ"),
			to:       []byte("---åäöÅÄÖ---"),
			in:       []byte(`a:1:{s:12:\"åäöÅÄÖ\";s:12:\"åäöÅÄÖ\";}`),
			out:      []byte(`a:1:{s:18:\"---åäöÅÄÖ---\";s:18:\"---åäöÅÄÖ---\";}`),
		},
```

Comment (abdullah-kasim) [reply search-replace_test.go:226]: These tests are working. Yeah not sure about adding it, it doesn't seem to add value because we don't require special handling for recursion unlike an actual serializer. So most of my tests are around byte calculation and escaping.

```suggestion
		},
		{
			testName: "exotic letters",
			from:     []byte("åäöÅÄÖ"),
			to:       []byte("---åäöÅÄÖ---"),
			in:       []byte(`a:1:{s:12:\"åäöÅÄÖ\";s:12:\"åäöÅÄÖ\";}`),
			out:      []byte(`a:1:{s:18:\"---åäöÅÄÖ---\";s:18:\"---åäöÅÄÖ---\";}`),
		},
		{
			testName: "complex array",
			from:     []byte("hallo"),
			to:       []byte("world!"),
			in:       []byte(`a:6:{i:0;i:1;i:1;d:1.1;i:2;s:5:\"hallo\";i:3;N;i:4;b:1;i:5;a:0:{}}`),
			out:      []byte(`a:6:{i:0;i:1;i:1;d:1.1;i:2;s:6:\"world!\";i:3;N;i:4;b:1;i:5;a:0:{}}`),
		},
		{
			testName: "object",
			from:     []byte("hello"),
			to:       []byte("world!"),
			in:       []byte(`O:8:\"stdClass\":1:{s:3:\"key\";s:5:\"hello\";}`),
			out:      []byte(`O:8:\"stdClass\":1:{s:3:\"key\";s:6:\"world!\";}`),
		},
		{
			testName: "object with recursion",
			from:     []byte("hello!"),
			to:       []byte("world!!!"),
			in:       []byte(`O:8:\"stdClass\":2:{s:4:\"obj2\";O:8:\"stdClass\":1:{s:4:\"obj3\";O:8:\"stdClass\":1:{s:4:\"obj1\";r:1;}}s:5:\"hello\";s:6:\"hello!\";}`),
			out:      []byte(`O:8:\"stdClass\":2:{s:4:\"obj2\";O:8:\"stdClass\":1:{s:4:\"obj3\";O:8:\"stdClass\":1:{s:4:\"obj1\";r:1;}}s:5:\"hello\";s:8:\"world!!!\";}`),
		},
		{
			testName: "replace object key with invalid value",
			from:     []byte("valid_key"),
			to:       []byte("invalid_key!!"),
			in:       []byte(`O:8:\"stdClass\":1:{s:9:\"valid_key\";s:4:\"data\";}`),
			out:      []byte(`O:8:\"stdClass\":1:{s:13:\"invalid_key!!\";s:4:\"data\";}`), //Should we replace the key or not?
		},
		{
			testName: "object with replace in key and value",
			from:     []byte("t1"),
			to:       []byte("t999"),
			in:       []byte(`O:8:\"stdClass\":1:{s:2:\"t1\";a:1:{s:2:\"t1\";a:1:{i:0;s:2:\"t1\";}}}`),
			out:      []byte(`O:8:\"stdClass\":1:{s:4:\"t999\";a:1:{s:4:\"t999\";a:1:{i:0;s:4:\"t999\";}}}`),
		},
```

File: search-replace.go [hunk 12]
```diff
@@ -129,6 +135,223 @@ func main() {
 	}
 }
 
+var debugMode = false
+
+func Debugf(format string, args ...interface{}) {
```
Comment (mjangda) [inline search-replace.go]: Do we want to keep the Debug function and calls? I had added this purely for development purposes.

File: search-replace.go [hunk 13]
```diff
@@ -129,6 +135,223 @@ func main() {
...
+			// index of the beginning of the next slice
+			nextSliceIndex = currentContentIndex + 3
+			// we're at backslash, so we need to minus 1 to get the index where the content finishes
+			contentEndIndex = currentContentIndex - 1
+			nextSliceFound = true
+			break
+		}
+
+		if contentByteCount > originalByteSize {
+			return nil, fmt.Errorf("faulty data, byte count does not match data size")
```
Comment (mjangda) [inline search-replace.go]: I would differentiate this error just a little bit so we know which path error-ed out.

File: search-replace.go (hunk 7 above)
Comment (mjangda) [reply search-replace.go]: Should we just remove the TODO and leave the comment as a note about the edge case?

**Review by mjangda — APPROVED** [review]
Shared some minor notes otherwise this looks good.

I did some separate performance testing as well (different size files, including some that are 20GB+, and have lots of serialized data). In most cases, this update is slower but not enough to be a concern. In one case (lots of serialized content), this was actually faster :) But I think results will vary depending on size, lines, complexity, etc.

File: search-replace.go (hunk 7 above)
Comment (abdullah-kasim) [reply search-replace.go]: Sure, that makes sense.

File: search-replace.go (hunk 12 above)
Comment (abdullah-kasim) [reply search-replace.go]: I'll remove it. Haven't found it to be too useful vs just breakpoints FWIW.
